File: kthSmallestElementInSortedMatrix.cpp
//time complexity:O(x+klogx)
//space complexity:O(x)
//executed on leetcode: yes
//approach:using min heap
//any issues faced? yes, very tricky

class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        priority_queue<pair<int,pair<int,int>>, vector<pair<int,pair<int,int>>>, greater<pair<int,pair<int,int>>>>q;
        int n=matrix.size();
        for(int i=0;i<n;i++)
        {
            q.push({matrix[0][i],{0,i}});
        }
        int ans;
        while(k--)
        {
            int val=q.top().first;
            int row=q.top().second.first;
            int col=q.top().second.second;
            q.pop();
            ans=val;
            if(row!=n-1)
            {
                q.push({matrix[row+1][col],{row+1,col}});
            }
        }
        return ans;
    }
};

File: kthsmallestelementInSortedMatrix.java
// Time Complexity : O(nlogn) 
// Space Complexity : O(l)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this: No 

class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        int min = matrix[0][0], max = matrix[n-1][n-1];
        
        while(min != max){
            int mid = min + (max - min) / 2;
            int count = countLessOrEqual(matrix, mid);
            if(count < k){
                min = mid + 1;
            }else{
                max = mid;
            }
        }
        return min;
    }
    public int countLessOrEqual(int [][] matrix, int mid){
        int count = 0;
        int column = matrix.length - 1;
        int row = 0;
        while(row < matrix.length && column >= 0){
            if(matrix[row][column] <= mid){
                count += column + 1;
                row++;
            } else{
                column--;
            }
        }
        return count;
    }
}